usim/_basics/resource.py

```diff
@@ -0,0 +1,208 @@
+from typing import TypeVar, Generic, Optional, Type
+
+from .._core.loop import __LOOP_STATE__
+from ._resource_level import __specialise__, ResourceLevels
+from .tracked import Tracked
+
+
+T = TypeVar('T')
+
+
+class BaseResources(Generic[T]):
+    """
+    Internal base class for resource types
+    """
+    _levels_type = None  # type: Type[ResourceLevels[T]]
+    _available = None  # type: Tracked[ResourceLevels[T]]
+
+    @property
+    def levels(self) -> ResourceLevels[T]:
+        """Current levels of resources"""
+        return self._available.value
+
+    @property
+    def resources_type(self) -> Type[ResourceLevels[T]]:
+        """Type of underlying resources"""
+        return self._levels_type
+
+    async def __insert_resources__(self, amounts: ResourceLevels):
+        new_levels = self._available.value + amounts
+        await self._available.set(new_levels)
+
+    async def __remove_resources__(self, amounts: ResourceLevels):
+        new_levels = self._available.value - amounts
+        await self._available.set(new_levels)
+
+    def borrow(self, **amounts: T) -> 'BorrowedResources[T]':
+        """
+        Temporarily borrow resources for a given context
+
+        :param amounts: resource levels to borrow
+        :return: async context to borrow resources
+        """
+        borrowed_levels = self._levels_type(**amounts)
+        assert self._levels_type.zero <= borrowed_levels,\
```

Comment:
Where is zero implemented at all?

Reply:
Zero is used very frequently for comparisons.